File: System-Programming-Fundamentals/lab5/part3/unpipe.c
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <zconf.h>

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s file\n", argv[0]);
        return EXIT_FAILURE;
    }

    int fd = open(argv[1], O_RDONLY);
    int pipe_fd[2];
    pipe(pipe_fd);
    int pipe_fd_rd = pipe_fd[0];
    int pipe_fd_wr = pipe_fd[1];

    int wc_pid = fork();
    if (wc_pid == 0) {
        //child process
        close(pipe_fd_wr);
        dup2(pipe_fd_rd, STDIN_FILENO);
        execlp("wc", "wc", NULL);
        close(pipe_fd_rd);
    } else {
        //parent process
        close(pipe_fd_rd);
        char file_char;
        while (read(fd, &file_char, 1) > 0) {
            write(pipe_fd_wr, &file_char, 1);
            read(fd, &file_char, 1); //skip one symbol
        }
        close(pipe_fd_wr);
    }
    close(fd);
    return EXIT_SUCCESS;
}

File: System-Programming-Fundamentals/lab5/part3/unpipe.pl
#!/usr/bin/perl -T
use strict;
use warnings qw(FATAL all);

$ENV{PATH} = '/usr/bin';
$ENV{ENV}="";

my $filename = shift @ARGV or die "Usage: $0 file\n";

open my $fd, '<', $filename or die "Can't open $filename!\n";
pipe my $pipe_fd_rd, my $pipe_fd_wr;

my $pid = fork();
if ($pid == 0) {
    close $pipe_fd_wr;
    open STDIN, '<&', $pipe_fd_rd;
    exec "wc";
}
else {
    close $pipe_fd_rd;
    my $file_char;
    while (read $fd, $file_char, 1) {
        print $pipe_fd_wr $file_char;
        read $fd, $file_char, 1;
    }
    close $pipe_fd_wr;
}
close($fd);